aggregator/src/aggregator.rs

```diff
@@ -1932,345 +1940,329 @@ impl VdafOps {
+                                    )
+                                }
+                                Ok((PingPongState::Finished(output_share), outgoing_message)) => (
+                                    ReportAggregationState::Finished,
                                     PrepareStepResult::Continue {
                                         message: outgoing_message,
                                     },
+                                    Some(output_share),
+                                ),
+                                Err(prepare_error) => (
+                                    ReportAggregationState::Failed { prepare_error },
+                                    PrepareStepResult::Reject(prepare_error),
                                     None,
-                                )
-                            }
-                            Ok((PingPongState::Finished(output_share), outgoing_message)) => (
-                                ReportAggregationState::Finished,
-                                PrepareStepResult::Continue {
-                                    message: outgoing_message,
-                                },
-                                Some(output_share),
-                            ),
-                            Err(prepare_error) => (
-                                ReportAggregationState::Failed { prepare_error },
-                                PrepareStepResult::Reject(prepare_error),
-                                None,
-                            ),
-                        };
-
-                    report_share_data.push(ReportShareData {
-                        report_share: prepare_init.report_share().clone(),
-                        report_aggregation: WritableReportAggregation::new(
-                            ReportAggregation::<SEED_SIZE, A>::new(
-                                *task.id(),
-                                aggregation_job_id,
-                                *prepare_init.report_share().metadata().id(),
-                                *prepare_init.report_share().metadata().time(),
-                                // A conversion failure here will fail the entire aggregation.
-                                // However, this is desirable: this can only happen if we receive
-                                // too many report shares in an aggregation job for us to store,
-                                // which is a whole-aggregation problem rather than a per-report
-                                // problem. (separately, this would require more than u64::MAX
-                                // report shares in a single aggregation job, which is practically
-                                // impossible.)
-                                ord.try_into()?,
-                                Some(PrepareResp::new(
+                                ),
+                            };
+
+                        sender.send(ReportShareData {
```

Comment:
I see that a SendError returned from here will cause `try_for_each_with()` to short-circuit, and stop other tasks as it can. This, combined with the fact that `send()` fails when the corresponding `UnboundedReceiver` is dropped, means that work being done in Rayon tasks will be stopped promptly when the initiating future is cancelled. We should describe this in a comment, maybe just above the Rayon parallel iterator calls.